src/Bases/Auctioneer/BaseAuctioneer.sol

```diff
@@ -0,0 +1,616 @@
+
+/**
+ *   @title Auctioneer Swaper
+ *   @author yearn.fi
+ *   @notice General use dutch auction contract for token sales.
+ */
+abstract contract BaseAuctioneer is BaseHealthCheck, ReentrancyGuard {
+    using SafeERC20 for ERC20;
+
+    /// @notice Emitted when a new auction is enabled
+    event AuctionEnabled(
+        bytes32 auctionId,
+        address indexed from,
+        address indexed to,
+        address indexed auctionAddress
+    );
+
+    /// @notice Emitted when an auction is disabled.
+    event AuctionDisabled(
+        bytes32 auctionId,
+        address indexed from,
+        address indexed to,
+        address indexed auctionAddress
+    );
+
+    /// @notice Emitted when auction has been kicked.
+    event AuctionKicked(bytes32 auctionId, uint256 available);
+
+    /// @notice Emitted when any amount of an active auction was taken.
+    event AuctionTaken(
+        bytes32 auctionId,
+        uint256 amountTaken,
+        uint256 amountLeft
+    );
+
+    /// @dev Store address and scaler in one slot.
+    struct TokenInfo {
+        address tokenAddress;
+        uint96 scaler;
+    }
+
+    /// @notice Store all the auction specific information.
+    struct AuctionInfo {
+        TokenInfo fromInfo;
+        uint96 kicked;
+        uint128 initialAvailable;
+        uint128 currentAvailable;
+    }
+
+    uint256 internal constant WAD = 1e18;
+
+    /// @notice Used for the price decay.
+    uint256 internal constant MINUTE_HALF_LIFE =
+        0.988514020352896135_356867505 * 1e27; // 0.5^(1/60)
+
+    /// @notice Struct to hold the info for `want`.
+    TokenInfo internal wantInfo;
+
+    /// @notice Mapping from an auction ID to its struct.
+    mapping(bytes32 => AuctionInfo) public auctions;
+
+    /// @notice Array of all the enabled auction for this contract.
+    bytes32[] public enabledAuctions;
+
+    /// @notice The amount to start the auction at.
+    uint256 public startingPrice;
+
+    /// @notice The time that each auction lasts.
+    uint32 public auctionLength;
+
+    /// @notice The minimum time to wait between auction 'kicks'.
+    uint32 public auctionCooldown;
+
+    /**
+     * @notice Initializes the Auction contract with initial parameters.
+     * @param _want Address this auction is selling to.
+     * @param _auctionLength Duration of each auction in seconds.
+     * @param _auctionCooldown Cooldown period between auctions in seconds.
+     * @param _startingPrice Starting price for each auction.
+     */
+    constructor(
+        address _asset,
+        string memory _name,
+        address _want,
+        uint32 _auctionLength,
+        uint32 _auctionCooldown,
+        uint256 _startingPrice
+    ) BaseHealthCheck(_asset, _name) {
+        require(auctionLength == 0, "initialized");
+        require(_want != address(0), "ZERO ADDRESS");
+        require(_auctionLength != 0, "length");
+        require(_auctionLength < _auctionCooldown, "cooldown");
```

Comment:
should we make this <= to allow for auctions that dont need any cooldown?

Reply:
maybe, i'm setting it to 1 day + 1 in my strategy to accomplish exactly this